encoding: implement PeekLength

RaduBerinde opened the pull request:
Implementing a function that figures out the length of a key-type encoding
without actually decoding the value. This will be used to parse keys and distsql
serialized values into EncDatums (see #6659).

@nvanbenschoten @paperstreet @andreimatei

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/6660)

<!-- Reviewable:end -->

paperstreet commented:
Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[util/encoding/encoding.go, line 859 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6660#-KHeJb2Xft_W8JOS9oVe:-KHeJb2Xft_W8JOS9oVf:13820541) ([raw file](https://github.com/cockroachdb/cockroach/blob/711dcf0ee79f21e01c9fe2d2053aa7884613e5c8/util/encoding/encoding.go#L859)):*
> ```Go
> 
> // PeekLength returns the length of the encoded value at the start of b. If the
> // encoded value is malformed, returns the length of the slice.
> ```

the length of the slice is an odd sentinal. why not an error?

---

*[util/encoding/encoding.go, line 884 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6660#-KHeKoq_PVw-EnlNefEN:-KHeKoq_PVw-EnlNefEO:1923955961) ([raw file](https://github.com/cockroachdb/cockroach/blob/711dcf0ee79f21e01c9fe2d2053aa7884613e5c8/util/encoding/encoding.go#L884)):*
> ```Go
> 		return 9
> 	}
> 	return getDecimalLen(b)
> ```

i see why you did this but i'm a little worried about it breaking non-obviously if we add something new

---

*[util/encoding/encoding_test.go, line 397 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6660#-KHeJmXfkQTveUTm_4vi:-KHeJmXfkQTveUTm_4vj:-1720866902) ([raw file](https://github.com/cockroachdb/cockroach/blob/711dcf0ee79f21e01c9fe2d2053aa7884613e5c8/util/encoding/encoding_test.go#L397)):*
> ```Go
> // that PeekLength returns the correct length.
> func testPeekLength(t *testing.T, encoded []byte) {
> 	// Append some garbage and test getDecimalLen again.
> ```

is this necessarily getDecimalLen?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6660)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[util/encoding/encoding.go, line 859 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6660#-KHeJb2Xft_W8JOS9oVe:-KHeqhLD_-uuAtICEHa3:356618256) ([raw file](https://github.com/cockroachdb/cockroach/blob/711dcf0ee79f21e01c9fe2d2053aa7884613e5c8/util/encoding/encoding.go#L859)):*
<details><summary><i>Previously, paperstreet (Dan Harrison) wrote…</i></summary>
> the length of the slice is an odd sentinal. why not an error?</details>
TFTR! My initial thought was that we would not detect all errors (since we don't actually decode everything) so we would still rely on decoding erroring out if something is malformed; and the code was easier to write this way. But in hindsight I agree those are flimsy excuses to have this fragile interface. I will switch to an error.

---

*[util/encoding/encoding.go, line 884 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6660#-KHeKoq_PVw-EnlNefEN:-KHeqfxwVvA94wxLOKpS:1020841676) ([raw file](https://github.com/cockroachdb/cockroach/blob/711dcf0ee79f21e01c9fe2d2053aa7884613e5c8/util/encoding/encoding.go#L884)):*
<details><summary><i>Previously, paperstreet (Dan Harrison) wrote…</i></summary>
> i see why you did this but i'm a little worried about it breaking non-obviously if we add something new</details>
will fix

---

*[util/encoding/encoding_test.go, line 397 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6660#-KHeJmXfkQTveUTm_4vi:-KHeqdqZ6z9BShAV1bE1:-1112293073) ([raw file](https://github.com/cockroachdb/cockroach/blob/711dcf0ee79f21e01c9fe2d2053aa7884613e5c8/util/encoding/encoding_test.go#L397)):*
<details><summary><i>Previously, paperstreet (Dan Harrison) wrote…</i></summary>
> is this necessarily getDecimalLen?</details>
copy/paste-o. i guess technically cut/paste-o :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6660)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Updated to return an error and cleaned up a little bit.

<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> #### encoding: implement PeekLength
> 
> Implementing a function that figures out the length of a key-type encoding
> 
> without actually decoding the value. This will be used to parse keys and distsql
> 
> serialized values into EncDatums (see #6659).
> 
> 
> 
> @nvanbenschoten @paperstreet @andreimatei
> 
> 
> </details>


---

Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6660#-:-KHez_5fop89vMJ7F6xm:1789895326)*
<!-- Sent from Reviewable.io -->

paperstreet commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Updated to return an error and cleaned up a little bit.</details>


---

Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[util/encoding/encoding.go, line 241 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6660#-KHf2OxgHV2w7eTleECh:-KHf2OxgHV2w7eTleECi:-391794858) ([raw file](https://github.com/cockroachdb/cockroach/blob/b687311d04a53c5664af15b2e8f3a9cd062a13ef/util/encoding/encoding.go#L241)):*
> ```Go
> 
> 	if length > len(b) {
> 		return 0, util.Errorf("varint length %d exceeds slice length %d",
> ```

nit: this wrapping is funky. pretty sure it'll fit on one line

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6660#-:-KHf3DOOMmCyEhqdK4DW:1417088721)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Reviewed 2 of 5 files at r1, 1 of 3 files at r2.
Review status: 2 of 5 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[util/encoding/decimal.go, line 367 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6660#-KHfDJmpoxo3mKb7g7oC:-KHfDJmpoxo3mKb7g7oD:1782136944) ([raw file](https://github.com/cockroachdb/cockroach/blob/711dcf0ee79f21e01c9fe2d2053aa7884613e5c8/util/encoding/decimal.go#L367)):*
> ```Go
> 		}
> 	}
> 	return -1, util.Errorf("did not find terminator %#x in buffer %#x", decimalTerminator, buf)
> ```

Please tell me this `util.Errorf` doesnt restrict the function's ability to be inlined. If it does, maybe just return the int and handle -1 caller-side. See `-gcflags="-m"`

---

*[util/encoding/encoding.go, line 483 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6660#-KHfJGwnMJwTduWEjY0l:-KHfJGwoWOOVNHGk75CL:-1277513217) ([raw file](https://github.com/cockroachdb/cockroach/blob/21e85ac9ff437de4a0b3cd3c31e0d6b1c33b4c72/util/encoding/encoding.go#L483)):*
> ```Go
> // escape, along with the value of the byte that follows.
> func findBytesEscape(b []byte, e escapes) (pos int, escapeVal byte, err error) {
> 	pos = bytes.IndexByte(b, e.escape)
> ```

Again, this might be paranoia, but do you mind just double checking that there were no benchmark regressions?

---

*[util/encoding/encoding.go, line 878 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6660#-KHfK7RPEOJmqso-9FI9:-KHfK7RQdPHO9g7f1goi:-119293119) ([raw file](https://github.com/cockroachdb/cockroach/blob/21e85ac9ff437de4a0b3cd3c31e0d6b1c33b4c72/util/encoding/encoding.go#L878)):*
> ```Go
> 	}
> 	m := b[0]
> 	switch m {
> ```

Are we missing encodings here? Where is int?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6660)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 1 of 5 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[util/encoding/decimal.go, line 367 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6660#-KHfDJmpoxo3mKb7g7oC:-KHfTPro7DYeKbiI2ITr:-277765315) ([raw file](https://github.com/cockroachdb/cockroach/blob/711dcf0ee79f21e01c9fe2d2053aa7884613e5c8/util/encoding/decimal.go#L367)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Please tell me this `util.Errorf` doesnt restrict the function's ability to be inlined. If it does, maybe just return the int and handle -1 caller-side. See `-gcflags="-m"`</details>
I checked and it looks like it doesn't get inlined even if I remove the Errorf. I didn't see a regression though, I think the function call cost is minor compared to the cost of decoding the decimals.

---

*[util/encoding/encoding.go, line 483 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6660#-KHfJGwnMJwTduWEjY0l:-KHfTKEG79YLERjPKUzx:-1006496058) ([raw file](https://github.com/cockroachdb/cockroach/blob/21e85ac9ff437de4a0b3cd3c31e0d6b1c33b4c72/util/encoding/encoding.go#L483)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Again, this might be paranoia, but do you mind just double checking that there were no benchmark regressions?</details>
Good catch. I did saw a 6-8% regression here, probably because it gets called in a loop. I reverted.

---

*[util/encoding/encoding.go, line 878 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6660#-KHfK7RPEOJmqso-9FI9:-KHfU26j70Q8bQfwr7JW:1359038532) ([raw file](https://github.com/cockroachdb/cockroach/blob/21e85ac9ff437de4a0b3cd3c31e0d6b1c33b4c72/util/encoding/encoding.go#L878)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Are we missing encodings here? Where is int?</details>
Oops, apparently I wasn't testing it for the ints, fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6660)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> And thanks for the benchmarks.

<details><summary><i>Previously, paperstreet (Dan Harrison) wrote…</i></summary>
> :lgtm:</details>


---

Review status: 1 of 5 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[util/encoding/decimal.go, line 367 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6660#-KHfDJmpoxo3mKb7g7oC:-KHf_YXctxxB28HZV7Hj:677042770) ([raw file](https://github.com/cockroachdb/cockroach/blob/711dcf0ee79f21e01c9fe2d2053aa7884613e5c8/util/encoding/decimal.go#L367)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> I checked and it looks like it doesn't get inlined even if I remove the Errorf. I didn't see a regression though, I think the function call cost is minor compared to the cost of decoding the decimals.</details>
SGTM

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6660#-:-KHf_jMqAJChejWk7i4Y:-2030648141)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
The crazy duration results were due to #6672.

<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> :lgtm: And thanks for the benchmarks.</details>


---

Review status: 1 of 5 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6660#-:-KHfeeVCm2TCLbQvENVo:65049053)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Ah, good catch!

<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> The crazy duration results were due to #6672.</details>


---

Review status: 1 of 5 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6660#-:-KHfiE4G6BVeqLsXOhcv:-480755959)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
encoding: implement PeekLength

Implementing a function that figures out the length of a key-type encoding
without actually decoding the value. This will be used to parse keys and distsql
serialized values into EncDatums.

Before/after benchmarks for decoding of decimals and bytes:
```
name                       old time/op  new time/op  delta
DecodeDecimal-4             839ns ±51%   652ns ±63%    ~      (p=0.222 n=9+9)
NonsortingDecodeDecimal-4   266ns ±10%   259ns ± 9%    ~      (p=0.367 n=8+8)
DecodeBytes-4              87.8ns ± 1%  88.7ns ± 1%  +1.04%  (p=0.004 n=7+10)
DecodeBytesDescending-4     107ns ± 0%   108ns ± 2%  +1.03%  (p=0.021 n=6+10)
```

Decode vs PeekLength benchmarks (note that for Bytes the decode times don't
include allocation of the output buffer):

```
name                            time/op
DecodeDecimal-4                 944ns ±145%
PeekLengthDecimal-4             33.5ns ± 2%

DecodeVarint-4                  29.0ns ± 1%
PeekLengthVarint-4              18.9ns ± 1%

DecodeUvarint-4                 25.3ns ± 1%
PeekLengthUvarint-4             18.9ns ± 1%

DecodeBytes-4                   87.3ns ± 1%
PeekLengthBytes-4               75.8ns ± 1%

DecodeBytesDescending-4          106ns ± 1%
PeekLengthBytesDescending-4     76.2ns ± 1%

DecodeDuration-4                2.71µs ±70%
PeekLengthDuration-4            1.99µs ±24%
```
